[Open in Colab](https://colab.research.google.com/github/WoolierBrooks/DementiaDetect-Speech/blob/main/scripts/adress_2020/00_Complete_nb.ipynb)

Importar bibliotecas

In [2]:
import pandas as pd
import numpy as np

Importar o dataset de treino

In [3]:
dataset = pd.read_csv("adress_20_db_emobase_train.csv")
X_train = dataset.iloc[:, :-1].values
y_train = dataset.iloc[:, -1].values
print(X_train, X_train.shape)
print(y_train, y_train.shape)

[[1.44e-05 4.17e-14 1.44e-05 ... 1.68e+00 1.35e+00 3.03e+00]
 [5.98e-06 4.17e-14 5.98e-06 ... 0.00e+00 0.00e+00 0.00e+00]
 [2.53e-05 0.00e+00 2.53e-05 ... 1.08e+00 9.27e-01 2.01e+00]
 ...
 [7.31e-05 0.00e+00 7.31e-05 ... 1.61e+00 1.64e+00 3.25e+00]
 [4.19e-05 5.43e-13 4.19e-05 ... 1.74e+00 1.78e+00 3.52e+00]
 [7.20e-05 3.34e-13 7.20e-05 ... 1.96e+00 1.72e+00 3.68e+00]] (108, 988)
[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1] (108,)


Importando o dataset de teste

In [5]:
dataset = pd.read_csv("adress_20_db_emobase_test.csv")
X_test = dataset.iloc[:, :-1].values
y_test = dataset.iloc[:, -1].values
print(X_test, X_test.shape)
print(y_test, y_test.shape)

[[6.10e-05 4.17e-14 6.10e-05 ... 1.41e+00 1.42e+00 2.83e+00]
 [2.30e-08 0.00e+00 2.30e-08 ... 0.00e+00 0.00e+00 0.00e+00]
 [3.15e-05 0.00e+00 3.15e-05 ... 1.17e+00 1.14e+00 2.31e+00]
 ...
 [2.00e-05 4.17e-14 2.00e-05 ... 5.32e-01 3.84e-01 9.16e-01]
 [4.40e-05 0.00e+00 4.40e-05 ... 1.30e+00 1.15e+00 2.45e+00]
 [1.58e-06 4.17e-14 1.58e-06 ... 1.16e+00 1.39e+00 2.56e+00]] (48, 988)
[0. 0. 1. 0. 1. 1. 0. 1. 1. 1. 0. 1. 0. 1. 0. 0. 1. 0. 0. 1. 0. 1. 1. 0.
 0. 1. 0. 1. 1. 1. 1. 1. 1. 0. 1. 0. 0. 0. 1. 0. 1. 0. 0. 1. 0. 1. 0. 0.] (48,)


Completando dados faltantes

In [6]:
from sklearn.impute import SimpleImputer

# Conjunto de treino
imputer = SimpleImputer(missing_values=np.nan, strategy="mean")
imputer.fit(X_train)
X_train = imputer.transform(X_train)
print(X_train, X_train.shape)

[[1.44e-05 4.17e-14 1.44e-05 ... 1.68e+00 1.35e+00 3.03e+00]
 [5.98e-06 4.17e-14 5.98e-06 ... 0.00e+00 0.00e+00 0.00e+00]
 [2.53e-05 0.00e+00 2.53e-05 ... 1.08e+00 9.27e-01 2.01e+00]
 ...
 [7.31e-05 0.00e+00 7.31e-05 ... 1.61e+00 1.64e+00 3.25e+00]
 [4.19e-05 5.43e-13 4.19e-05 ... 1.74e+00 1.78e+00 3.52e+00]
 [7.20e-05 3.34e-13 7.20e-05 ... 1.96e+00 1.72e+00 3.68e+00]] (108, 988)


In [7]:
# Conjunto de teste
imputer = SimpleImputer(missing_values=np.nan, strategy="mean")
imputer.fit(X_test)
X_test = imputer.transform(X_test)
print(X_test, X_test.shape)

[[6.10e-05 4.17e-14 6.10e-05 ... 1.41e+00 1.42e+00 2.83e+00]
 [2.30e-08 0.00e+00 2.30e-08 ... 0.00e+00 0.00e+00 0.00e+00]
 [3.15e-05 0.00e+00 3.15e-05 ... 1.17e+00 1.14e+00 2.31e+00]
 ...
 [2.00e-05 4.17e-14 2.00e-05 ... 5.32e-01 3.84e-01 9.16e-01]
 [4.40e-05 0.00e+00 4.40e-05 ... 1.30e+00 1.15e+00 2.45e+00]
 [1.58e-06 4.17e-14 1.58e-06 ... 1.16e+00 1.39e+00 2.56e+00]] (48, 988)


Escalonamento de características

In [8]:
from sklearn.preprocessing import StandardScaler

sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)
print(X_train, X_train.shape)
print(X_test, X_test.shape)

[[-0.39577152 -0.21254338 -0.39577152 ...  0.70402515  0.40177223
   0.55626519]
 [-0.55834326 -0.21254338 -0.55834326 ... -1.03157583 -1.01495572
  -1.02738984]
 [-0.18531641 -0.23891375 -0.18531641 ...  0.08416766 -0.04213586
   0.02315359]
 ...
 [ 0.7375968  -0.23891375  0.7375968  ...  0.63170844  0.70610639
   0.67125004]
 [ 0.1351932   0.10447027  0.1351932  ...  0.7660109   0.85302632
   0.81236781]
 [ 0.71635821 -0.02769779  0.71635821 ...  0.99329198  0.79006064
   0.89599316]] (108, 988)
[[ 0.50397233 -0.21254338  0.50397233 ...  0.42508928  0.4752322
   0.4517335 ]
 [-0.67335987 -0.23891375 -0.67335987 ... -1.03157583 -1.01495572
  -1.02738984]
 [-0.065608   -0.23891375 -0.065608   ...  0.17714628  0.18139233
   0.17995112]
 ...
 [-0.28764779 -0.21254338 -0.28764779 ... -0.48196885 -0.61197533
  -0.54863472]
 [ 0.17573959 -0.23891375  0.17573959 ...  0.31144874  0.19188661
   0.2531233 ]
 [-0.64329761 -0.21254338 -0.64329761 ...  0.16681532  0.44374936
   0.31061573]] (48, 9

Testando Linear Discriminant Analysis (LDA)

In [21]:
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.metrics import accuracy_score

# Instanciando o classificador LDA
lda = LinearDiscriminantAnalysis()

# Treinando o modelo LDA com os dados de treinamento
lda.fit(X_train, y_train)

# Realizando previsões nos dados de teste
y_pred = lda.predict(X_test)

# Calculando a precisão da classificação
accuracy = accuracy_score(y_test, y_pred)
print(f"Acurácia do modelo: {accuracy * 100:.2f}%")

Acurácia do modelo: 60.42%


Testando Decision Tree (DT)

In [18]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

# Inicializa o classificador de árvore de decisão
clf = DecisionTreeClassifier(random_state=42)

# Treina o classificador com o conjunto de treino
clf.fit(X_train, y_train)

# Realiza a previsão com o conjunto de teste
y_pred = clf.predict(X_test)

# Calcula a acurácia do modelo
accuracy = accuracy_score(y_test, y_pred)
print(f'Acurácia do modelo: {accuracy * 100:.2f}%')


Acurácia do modelo: 41.67%


Testando Rede Neural Simples (INN)

In [20]:
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import accuracy_score

# Criando o classificador da rede neural
classifier = MLPClassifier(hidden_layer_sizes=(10,), max_iter=1000, random_state=42)

# Treinando o classificador
classifier.fit(X_train, y_train)

# Realizando previsões no conjunto de teste
predictions = classifier.predict(X_test)

# Calculando a acurácia do modelo
accuracy = accuracy_score(y_test, predictions)
print(f'Acurácia do modelo: {accuracy * 100:.2f}%')


Acurácia do modelo: 66.67%


Testando Support Vector Machine (SVM)

In [29]:
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score

# Criar um classificador SVM
classificador_svm = SVC(kernel='linear')  # Você pode escolher o tipo de kernel desejado

# Treinar o classificador SVM
classificador_svm.fit(X_train, y_train)

# Fazer previsões usando o conjunto de teste
predictions = classificador_svm.predict(X_test)

# Calculando a acurácia do modelo
accuracy = accuracy_score(y_test, predictions)
print(f'Acurácia do modelo: {accuracy * 100:.2f}%')

Acurácia do modelo: 72.92%


Testando Random Forest (RF)

In [36]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report

# Criando o classificador Random Forest
classificador_rf = RandomForestClassifier(n_estimators=100, random_state=42)

# Treinando o classificador com os dados de treinamento
classificador_rf.fit(X_train, y_train)

# Fazendo previsões com o conjunto de teste
predictions = classificador_rf.predict(X_test)

# Calculando a acurácia do modelo
accuracy = accuracy_score(y_test, predictions)
print(f'Acurácia do modelo: {accuracy * 100:.2f}%')

Acurácia do modelo: 62.50%


In [37]:
print(X_train, X_train.shape)
print(y_train, y_train.shape)

[[-0.39577152 -0.21254338 -0.39577152 ...  0.70402515  0.40177223
   0.55626519]
 [-0.55834326 -0.21254338 -0.55834326 ... -1.03157583 -1.01495572
  -1.02738984]
 [-0.18531641 -0.23891375 -0.18531641 ...  0.08416766 -0.04213586
   0.02315359]
 ...
 [ 0.7375968  -0.23891375  0.7375968  ...  0.63170844  0.70610639
   0.67125004]
 [ 0.1351932   0.10447027  0.1351932  ...  0.7660109   0.85302632
   0.81236781]
 [ 0.71635821 -0.02769779  0.71635821 ...  0.99329198  0.79006064
   0.89599316]] (108, 988)
[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1] (108,)


Resultados obtidos:
|Características|LDA   |DT    |INN   |SVM   |RF    |Média |
|:-------------:|:----:|:----:|:----:|:----:|:----:|:----:|
|emobase (988)  |60.42%|41.67%|66.67%|72.92%|62.50%|60.84%|